Add March 2021 Infrastructure Changelog (#1)
ViktorHofer/runtime: docs/infra/changelog.md

# March 2021

> Welcome to the March 2021 changelog of dotnet/runtime's infrastructure. There are a number of updates that we hope you will like, some of the key highlights include:

Reviewer: how often will they see this mailer? 
what will they get out of it -- why should they not stop reading?

- :heavy_check_mark: [Enable building runtime repository on Windows ARM64](#enable-building-runtime-repository-on-windows-arm64)
- :heavy_check_mark: [Creating a separate pipeline to stage onboarding new test targets](#creating-a-separate-pipeline-to-stage-onboarding-new-test-targets)

### :bulb: Enhancements
- [Consolidate VS detection logic](https://github.com/dotnet/runtime/pull/49593)
- [Make sure event generation is incremental](https://github.com/dotnet/runtime/pull/48903)
- [Move dotnet-sdk snaps into the installer repo and dotnet-runtime snaps into servicing branches](https://github.com/dotnet/runtime/issues/49374)
- [Port Microsoft.Extensions.* from pkgproj to dotnet pack](https://github.com/dotnet/runtime/pull/48385)
- [Update debug dump template after runfo update to include dumps](https://github.com/dotnet/runtime/pull/49183)
- [Update the minimum version of the SDK to 6.0 Preview 2](https://github.com/dotnet/runtime/pull/50084)
- [Reorganize CoreCLR native build to reduce CMake reconfigures when the build system is untouched](https://github.com/dotnet/runtime/pull/49906)

> ### :x: Bug fixes

Reviewer: I would use a different icon (a bug maybe?) since the X looks like they're new bugs, whereas you actually fixed them..

> ### :x: Bug fixes

Reviewer: ```suggestion
### :bug: New fixes
```
eg

> - [Code coverage is broken](https://github.com/dotnet/runtime/issues/49172)

Reviewer: When fixing something that broke, it might be worth a note explaining how it broke (so it doesn't feel like infrastructure is always on a treadmill of breaking and fixing things 😄 . In this case, looks like we had to work around an SDK bug.

- [Error when building using the -ninja flag](https://github.com/dotnet/runtime/issues/47314)
- [Unable to build repo on latest VS dogfood](https://github.com/dotnet/runtime/issues/49646)
- [The repo fails to build when source control information isn't available](https://github.com/dotnet/runtime/issues/47130)
- [Trimming tests fail with race condition](https://github.com/dotnet/runtime/issues/40398)

## Unified GitHub project board

> Since consolidation of repositories into dotnet/runtime, we used to have different infrastructure boards to track work. We recently consolidated those into a [single Infrastructure board](https://github.com/dotnet/runtime/projects/26) which is kept up-to-date via webhooks.

Reviewer: History isn't "useful". 
```suggestion
We now track all infrastructure work on a [single board](https://github.com/dotnet/runtime/projects/26) and we keep it up to date. You can go there to see what we are working on and what is coming up.
```

The consolidated board allows to
- Keep the number of untriaged issues low
- Track Servicing, 6.0 and Future work

> - Summarize accomplishments ("Done" column) over the last month which ultimately results in this changelog

Reviewer: is this useful to share?

> As you can see in the image below, the number of infrastructure issues is quite high. We will start triaging these more aggressively starting soon.

Reviewer: ```suggestion
As you can see in the image below, the number of infrastructure issues is quite high. Before the next update, we will triage these aggressively.
```

"soon" is "weasel words" 😁

![image](https://user-images.githubusercontent.com/7412651/112136833-cd130900-8bcf-11eb-95a3-d632fcafa496.png)

## Default branch has been renamed to main

> The default branch in the runtime, runtime-assets, runtimelab, jitutils and other repositories that we own [has been renamed to main](https://github.com/dotnet/runtime/issues/48357).

Reviewer: ```suggestion
We [renamed the default branch to main](https://github.com/dotnet/runtime/issues/48357) runtime, runtime-assets, runtimelab, jitutils and other repos we own.
```

> Besides a call to action to rename forked repositories' default branch to main, there wasn't any noticeable impact. Some team members even reached out to us and expressed their gratitude.

Reviewer: ```suggestion
Besides a call to action to rename forked repositories' default branch to main, there wasn't any noticeable impact.
```
avoid self-aggrandizing 🙂

> :memo: If you haven't renamed your fork's default branch to main, we strongly advise to do so.

Reviewer: link to instructions?

> ## Improve CoreCLR Windows native build time

Reviewer: Use active verbs/phrasing. eg.,
```suggestion
## CoreCLR build is now up to 50% faster on Windows
```

> Building the CoreClr runtime on Windows was reported as unsatisfying, as the build takes quite long (up to 30 minutes depending on the hardware used) and it makes the machine unusable during that time.

Reviewer: consider recasting your work in terms of addressing the readers' needs. that's the only reason they're reading this. eg., in this case maybe
```suggestion
You told us that building the CoreClr runtime on Windows was was too slow. It could take up to 30 minutes and make the machine unusable during that time.
```

> MSBuild's support for effective parallelisation is limited. To make builds as fast as possible, tools like Ninja are recommended to be used. Because of that, support for Ninja over MSBuild was added months ago in the runtime repository via the "-ninja" build switch. Depending on the hardware being used, the improvements vary between 10-50%.

Reviewer: Active voice. Short useful sentences.

```suggestion
You can increase the speed of the coreclr build by enabling a tool called Ninja. It handles parallelism better than MSBuild. You can enable it by adding the "-ninja" switch to your build command. Your build will speed up 10-50%.
```

Why would they not want to add this switch? why is it not default?

Reviewer: Oh, I just realized they don't need the flag anymore. I think you can rephrase this in terms of, you don't need this flag any more. You will automatically see the faster builds.

:bulb: What just recently changed is that Ninja is now the [default CMake generator on Windows](https://github.com/dotnet/runtime/pull/49715).

> ## Enable building runtime repository on Windows ARM64

Reviewer: ```suggestion
## You can now build the repo on a Windows ARM64 box
```
etc. cast in terms of "what you got". aka "why should I read this"

> The runtime repository failed to build on Windows ARM64 devices like the Surface Pro X, mainly because the wrong set of dac tools were chosen. This was  [just recently fixed](https://github.com/dotnet/runtime/pull/49864).

Reviewer: Is all this detail necessary?  The minority that want to see the details can click a link
```suggestion
With a recent [fix](https://github.com/dotnet/runtime/pull/49864) you can build the runtime repo on a Windows ARM64 device like a Surface Pro X.
```

:warning: The x86 emulation of Git for Windows doesn't work yet on Windows ARM64. Native ARM64 execution is expected to be official supported in one of the following month's releases. See https://github.com/git-for-windows/git/issues/3107 for more details.

## Creating a separate pipeline to stage onboarding new test targets

> Even though the [runtime-staging pipeline](https://dnceng.visualstudio.com/public/_build?definitionId=924) has already been running as part of the CI for serveral months, we never officially announced it.

Reviewer: delete paragraph? no need for "narrative"

> In the past, when new configurations (i.e. Windows ARM64 or WASM) were added to the repository, they often regressed reliability of the main runtime pipeline. In addition to that, we often encounter hardware issues (i.e. limited number of devices, malfunctioning hardware, bad rollouts) which require the flexibility to move configurations out of the main runtime pipeline and "quarantine" them into a separate pipeline.

Reviewer: As a general rule, avoid "passive voice" in communication. Basically change from abstract to I/you/us. It makes it more engaging and effective. 

Eg., "which require the flexibility" --> "which means we have to"/"which needs us to be able to"

```suggestion
We created the staging pipeline so help prove out and stabilize new configurations like Windows ARM64 and WASM before we add them to the of the main pipeline. It's also a place we can move configurations while they work through hardware problems like a shortage of working devices. This "quarantine" system will help avoid these problems impacting PR validation and official builds.
```

BTW I think in this specific case, I would also make clear how this will be presented to them, eg., will they see these in PR validations? And if so, how do htey know to ignore them?

For that purpose the runtime-staging pipeline was added to the repository and runs as part of every Pull Request and Rolling Build.

:information_source: Test failures happening in that pipeline don't impact the overall status of the PR.

> :warning: Configurations ("legs") in that pipeline which time out (i.e. because of limited devices being available) count as failures and impact the Pull Request's status.

Reviewer: What action do you need folks to take in this case? What are they supposed to do?
